File: codes/string/151./reverse-words-in-a-string/reverse-words-in-a-string.go
package main

import "bytes"
func reverseWords(s string) string {
	var newS bytes.Buffer
	i:=len(s)-1
	j:=i
	for i>=0{
		for i>=0&&s[i]==' '{
			i--
		}
		j=i
		for j>=0&&s[j]!=' '{
			j--;
		}
		if j<0{
			newS.WriteString(s[0:i+1])
		}else{
			newS.WriteString(s[j+1:i+1])
			newS.WriteRune(' ')
		}
        i=j-1;
	}
	rs:=newS.String()
	if rs[len(rs)-1]==' '{
		return rs[:len(rs)-1]
	}else{
		return rs
	}

}

File: codes/string/151./reverse-words-in-a-string/reverse-words-in-a-string.cpp
#include <string>
#include <stack>
using namespace std;
class Solution {
public:
    string reverseWords(string s) {
        string newS;
        for (int i = s.size() - 1; i >= 0;) {
            while(i>=0&&s[i]==' '){
                i--;
            }
            int j = i;
            while(j>=0&&s[j]!=' '){
                j--;
            }
            if(j<0){
                newS.append(s.substr(0, i+1));
            }else{
                newS.append(s.substr(j + 1, i - j));
                newS.append(" ");
            }
            i = j;
        }
        if(*(newS.end()-1)==' '){
            newS.erase(newS.end() - 1);
        }

        return newS;
    }
};